mono/metadata/threads.c

```diff
@@ -234,154 +235,110 @@ mono_thread_get_abort_prot_block_count (MonoInternalThread *thread)
-			} else {
-				new_state |= INTERRUPT_REQUEST_DEFERRED_BIT;
-			}
-		}
+	/* pending exception, self abort */
+	if (state & INTERRUPT_SYNC_REQUESTED_BIT)
+		return TRUE;
 
-		//bounds check abort_prot_count
-		g_assert (new_val >= 0);
-		g_assert (new_val < (1 << ABORT_PROT_BLOCK_BITS));
-		new_state |= new_val << ABORT_PROT_BLOCK_SHIFT;
+	/* abort, interruption, suspend */
+	if ((state & INTERRUPT_ASYNC_REQUESTED_BIT) && !(state & ABORT_PROT_BLOCK_MASK))
+		return TRUE;
 
-	} while (InterlockedCompareExchangePointer ((volatile gpointer)&thread->thread_state, (gpointer)new_state, (gpointer)old_state) != (gpointer)old_state);
-	return (new_state & INTERRUPT_REQUESTED_BIT) == INTERRUPT_REQUESTED_BIT;
+	return FALSE;
 }
 
-
-//Don't use this function, use inc/dec below
-static void
-mono_thread_abort_prot_block_count_add (MonoInternalThread *thread, int val)
+gboolean
+mono_threads_end_abort_protected_block (void)
 {
+	MonoInternalThread *thread = mono_thread_internal_current ();
 	gsize old_state, new_state;
 	do {
 		old_state = thread->thread_state;
-		verify_thread_state (old_state);
 
-		int new_val = val + ((old_state & ABORT_PROT_BLOCK_MASK) >> ABORT_PROT_BLOCK_SHIFT);
 		//bounds check abort_prot_count
+		int new_val = ((old_state & ABORT_PROT_BLOCK_MASK) >> ABORT_PROT_BLOCK_SHIFT) - 1;
 		g_assert (new_val >= 0);
 		g_assert (new_val < (1 << ABORT_PROT_BLOCK_BITS));
-		new_state = (old_state & ~ABORT_PROT_BLOCK_MASK) | (new_val << ABORT_PROT_BLOCK_SHIFT);
 
+		new_state = old_state - (1 << ABORT_PROT_BLOCK_SHIFT);
 	} while (InterlockedCompareExchangePointer ((volatile gpointer)&thread->thread_state, (gpointer)new_state, (gpointer)old_state) != (gpointer)old_state);
-}
 
-static void
-mono_thread_inc_abort_prot_block_count (MonoInternalThread *thread)
-{
-	mono_thread_abort_prot_block_count_add (thread, 1);
-}
-
-static void
-mono_thread_dec_abort_prot_block_count (MonoInternalThread *thread)
-{
-	mono_thread_abort_prot_block_count_add (thread, -1);
+	return mono_thread_state_has_interruption (new_state);
 }
 
 static gboolean
 mono_thread_get_interruption_requested (MonoInternalThread *thread)
 {
 	gsize state = thread->thread_state;
-	return (state & INTERRUPT_REQUESTED_BIT) == INTERRUPT_REQUESTED_BIT;
+
+	return mono_thread_state_has_interruption (state);
 }
 
-/* Returns TRUE is there was a state change */
+/*
+ * Returns TRUE is there was a state change
+ * We clear a single interruption request, sync has priority.
+ */
 static gboolean
 mono_thread_clear_interruption_requested (MonoInternalThread *thread)
 {
 	gsize old_state, new_state;
 	do {
 		old_state = thread->thread_state;
-		verify_thread_state (old_state);
 
 		//Already cleared
-		if (!(old_state & (INTERRUPT_REQUESTED_BIT | INTERRUPT_REQUEST_DEFERRED_BIT)))
+		if (!(old_state & INTERRUPT_REQUESTED_MASK))
 			return FALSE;
-		new_state = old_state & ~(INTERRUPT_REQUESTED_BIT | INTERRUPT_REQUEST_DEFERRED_BIT);
+
+		if (old_state & INTERRUPT_SYNC_REQUESTED_BIT)
```

Comment:
This design remains unsound in my head. So, first, let me phrase all the scenarios and things to consider.

There are 3 sources of abort we need to consider. set_pending, sync abort and async abort.
There are abort protected regions that delay the execution of aborts.

set_pending and sync abort must be executed unconditionally.
async aborts must be delayed around protected regions.

Can we have multiple pending aborts? Yes, we can have set_pending happen in tandem of an async abort. In case of a protected block, we must deliver only set_pending exception while keeping the async one deferred.

Looks like your change does what I think needs done.

Hmmm, so, in addition to the above fix, we need a couple of extra things.

- debug/trace level logging of those things. Logging thread state and thread_interruption_requested will make it easier to spot issues like unbalanced bumps and threads going into bad states.

- We lost some asserts for valid states.